store/src/store.rs

```diff
@@ -12,168 +14,17 @@ use ckb_core::transaction::{CellOutPoint, ProposalShortId, Transaction};
 use ckb_core::transaction_meta::TransactionMeta;
 use ckb_core::uncle::UncleBlock;
 use ckb_core::{Bytes, EpochNumber};
-use ckb_db::{Col, DbBatch, Error, KeyValueDB};
-use ckb_protos::{self as protos, CanBuild};
-use lru_cache::LruCache;
+use ckb_db::Col;
+use ckb_protos as protos;
 use numext_fixed_hash::H256;
-use serde_derive::{Deserialize, Serialize};
 use std::convert::TryInto;
-use std::ops::Range;
-use std::sync::Mutex;
 
-const META_TIP_HEADER_KEY: &[u8] = b"TIP_HEADER";
-const META_CURRENT_EPOCH_KEY: &[u8] = b"CURRENT_EPOCH";
-
-#[derive(Clone, Serialize, Deserialize, Eq, PartialEq, Hash, Debug)]
-pub struct StoreConfig {
-    pub header_cache_size: usize,
-    pub cell_data_cache_size: usize,
-}
-
-impl Default for StoreConfig {
-    fn default() -> Self {
-        Self {
-            header_cache_size: 4096,
-            cell_data_cache_size: 128,
-        }
-    }
-}
-
-pub struct ChainKVStore<T> {
-    db: T,
-    header_cache: Mutex<LruCache<H256, Header>>,
-    cell_data_cache: Mutex<LruCache<(H256, u32), Bytes>>,
-}
-
-impl<T: KeyValueDB> ChainKVStore<T> {
-    pub fn new(db: T) -> Self {
-        Self::with_config(db, StoreConfig::default())
-    }
-
-    pub fn with_config(db: T, config: StoreConfig) -> Self {
-        ChainKVStore {
-            db,
-            header_cache: Mutex::new(LruCache::new(config.header_cache_size)),
-            cell_data_cache: Mutex::new(LruCache::new(config.cell_data_cache_size)),
-        }
-    }
-
-    pub fn get(&self, col: Col, key: &[u8]) -> Option<Vec<u8>> {
-        self.db.read(col, key).expect("db operation should be ok")
-    }
-
-    pub fn exists(&self, col: Col, key: &[u8]) -> bool {
-        self.get(col, key).is_some()
-    }
-
-    pub fn partial_get(&self, col: Col, key: &[u8], range: &Range<usize>) -> Option<Vec<u8>> {
-        self.db
-            .partial_read(col, key, range)
-            .expect("db operation should be ok")
-    }
-
-    fn process_get<F, Ret>(&self, col: Col, key: &[u8], process: F) -> Option<Ret>
-    where
-        F: FnOnce(&[u8]) -> Result<Option<Ret>, Error>,
-    {
-        self.db
-            .process_read(col, key, process)
-            .expect("db operation should be ok")
-    }
-
-    pub fn traverse<F>(&self, col: Col, callback: F) -> Result<(), Error>
-    where
-        F: FnMut(&[u8], &[u8]) -> Result<(), Error>,
-    {
-        self.db.traverse(col, callback)
-    }
-}
-
-/// Store interface by chain
-pub trait ChainStore: Sync + Send {
-    /// Batch handle
-    type Batch: StoreBatch;
-    /// New a store batch handle
-    fn new_batch(&self) -> Result<Self::Batch, Error>;
+pub trait ChainStore<'a>: Send + Sync {
+    type Vector: AsRef<[u8]>;
+    fn get(&'a self, col: Col, key: &[u8]) -> Option<Self::Vector>;
```

Comment:
@yangby-cryptape added a `process_get` and changed all db read fn with closure in this PR:

https://github.com/nervosnetwork/ckb/pull/1032/files#diff-65e0871e65c7b9fa5eb3827ae3a6a9b5R71

May I know the reason to change it back?

Reply:
`process_get` take closure to deal with lifetime when use `get_pinned`, It does not have to take closure now, since lifetime already bind with trait `ChainStore`.

Reply:
Get, thanks.